Enums with generics are incompatible with this macro

The following code

```rust
#[derive(EnumAsInner)]
enum Test<T> {
    A(T),
}
```
fails to compile because

```
error[E0412]: cannot find type `T` in this scope
 --> src\main.rs:5:7
  |
5 |     A(T),
  |       ^ not found in this scope

error[E0107]: wrong number of type arguments: expected 1, found 0
 --> src\main.rs:4:6
  |
4 | enum Test<T> {
  |      ^^^^ expected 1 type argument

error: aborting due to 2 previous errors

```

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -294,6 +294,7 @@ fn named_fields_return(
 
 fn impl_all_as_fns(ast: &DeriveInput) -> TokenStream {
     let name = &ast.ident;
+    let generics = &ast.generics;
 
     let enum_data = if let syn::Data::Enum(data) = &ast.data {
         data
@@ -363,8 +364,10 @@ fn impl_all_as_fns(ast: &DeriveInput) -> TokenStream {
         stream.extend(tokens);
     }
 
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
     quote!(
-        impl #name {
+        impl #impl_generics #name #ty_generics #where_clause {
             #stream
         }
     )

--- a/tests/generics.rs
+++ b/tests/generics.rs
@@ -1,0 +1,24 @@
+use enum_as_inner::EnumAsInner;
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Copy, EnumAsInner)]
+enum WithGenerics<T: Clone + Copy> {
+    A(T),
+    B(T),
+}
+
+#[test]
+fn with_generics() {
+    let mut with_generics = WithGenerics::A(100);
+
+    assert!(with_generics.as_a().is_some());
+    assert!(with_generics.as_b().is_none());
+
+    assert_eq!(with_generics.into_a().unwrap(), 100);
+    assert_eq!(*with_generics.as_a().unwrap(), 100);
+    assert_eq!(*with_generics.as_a_mut().unwrap(), 100);
+
+    assert!(with_generics.into_b().is_err());
+    assert!(with_generics.as_b().is_none());
+    assert!(with_generics.as_b_mut().is_none());
+}

